### ISSUE TITLE
Constant crashes on Manjaro KDE

### ISSUE THREAD

**Issue by Memory383** [xiaoyifang/goldendict#371]
I did a fresh installation on Manjaro KDE Stable via AUR Repository on March 5th.

**There are 2 issues that were not present 2 months ago:**

1) GUI does not show properly after running the software. The dictionary groups are not shown. The user cannot look for any word:

![grafik](https://user-images.githubusercontent.com/121114125/223046225-328164aa-191a-4b56-b98d-638694191132.png)

GoldenDict loads the GUI correctly after several minutes.

2) Constant crashes. The software sometimes does not load at all. 

-------------
Version info:

Goldendict 23.02.07-alpha.2404a5d3 on Sun Mar 5 13:56:47 2023
manjaro linux 5.15.94-1-MANJARO Qt 5.15.8 x86_64-little_endian-lp64
GCC 12.2.1 20230201
Flags: MAKE_ZIM_SUPPORT MAKE_EXTRA_TIFF_HANDLER MAKE_CHINESE_CONVERSION_SUPPORT

**Comment by shenlebantongying** [comment]
Can you check your systemd's coredump service and send a back trace? 
```
coredumpctl list # list all coredumps, notice the PID row then get a recent one about goldendict
coredumpctl info [PID]
```
https://wiki.archlinux.org/title/Core_dump#Examining_a_core_dump

Also, can you check if the program still crashes without any dicts or without recent dicts?

**Comment by shenlebantongying** [comment]
I think the crashes is related to a clipboard change. All the traces contains `_ZN10QClipboard7changedENS_4ModeE` (with name mangling → QClipboard changed). Maybe something you copied will cause a crash? 
```
Stack trace of thread 1189:
#0  0x00005600111f99b8 n/a (goldendict + 0xea9b8)
#1  0x00007f4d84b69a71 n/a (libQt5Core.so.5 + 0x2bea71)
#2  0x00007f4d859f54c3 _ZN10QClipboard7changedENS_4ModeE (libQt5Gui.so.5 + 0x5184c3)
#3  0x00007f4d72efbc2b _ZN14QXcbConnection14handleXcbEventEP19xcb_generic_event_t (libQt5XcbQpa.so.5 + 0x3cc2b)
#4  0x00007f4d72eff4c0 _ZN14QXcbConnection16processXcbEventsE6QFlagsIN10QEventLoop17ProcessEventsFlagEE (libQt5XcbQpa.so.5 + 0x404c0)
#5  0x00007f4d72f238c8 n/a (libQt5XcbQpa.so.5 + 0x648c8)
```

**Comment by Memory383** [comment]
@shenlebantongying  :  Yes, the crashes on GoldenDict might be related to the clipboard. _The problem seems a regression introduced with the commits from the last 2 months._ 

I have not changed the OS neither the software on my PC. Only GoldenDict was updated and now crashes.

**Let's keep this issue open to see if someone else has the same problem.**

**Comment by xiaoyifang** [comment]
maybe related with [b3aac01f41d523289945f5c32e0ce573a6e3b133](https://github.com/xiaoyifang/goldendict/commit/b3aac01f41d523289945f5c32e0ce573a6e3b133)

in the clipboardchange event,  the scanpop is null and caused a segv error

**Comment by shenlebantongying** [comment]
but there is no code that destroy/reset scanpopup anymore. Setting parent to nullptr will make it live forever (the original code also set it to nullptr). I am not sure.

**Comment by xiaoyifang** [comment]
![image](https://user-images.githubusercontent.com/105986/223294248-56fa1452-8b4d-404e-88c9-870b75cb0dfa.png)

As it is difficult to reproduced the crash.
we can have some test, 
method 1:
   add the `if condition` back in a new commit and have a try.
or 
method 2: revert the commit and have a try.

If the scanpopup is not the root cause ,I can reset the commit history and changed to current commit history state.

**Comment by xiaoyifang** [comment]
https://github.com/xiaoyifang/goldendict/blob/staged/mainwindow.cc#L266-L280


![image](https://user-images.githubusercontent.com/105986/223299536-3cfefce3-56de-4ef9-a6a5-86c2b7e7fe64.png)

**Comment by xiaoyifang** [comment]
@Memory383 I add some conditon checks ,can you try again .

**Comment by shenlebantongying** [comment]
I don't know. Without clear reproduce steps, I don't want to fix anything because it does not crash on my machine.

**Comment by xiaoyifang** [comment]
I guess the reproduce step can be like this:
1, check enable scanpopup when start
2. exit the goldendict
3. start the goldendict again.

**Comment by xiaoyifang** [comment]
> the original code also set it to nullptr

some note on this, the original code use shared_ptr to manage the memory release ,which makes it's ok to pass the nullptr.
when changed into c++ pointer ,we have to release the memory by ourselves. In this case ,we have to pass `this` (managed by qt )or delete the pointer at MainWindows's destruction function.

**Comment by shenlebantongying** [comment]
That change doesn't matter. The trigger() only connect QClipboard::changed and Mainwindow::clipbaord_changed. The clipboard_changed's code won't be executed unless clipboard is changed before scanpop's initialization. The timing between them are less than a few milliseconds, the chance is **extremely** rare.

Even with this extremely rare problem, the GD will start normally without crash on my machine.

This bug report is related to something else.

**Comment by shenlebantongying** [comment]
`this` and nullptr is not related to possible crashes too, because Scanpopup won't be destroyed at all unless the app quit.

**Comment by xiaoyifang** [comment]
> The trigger() only connect QClipboard::changed and Mainwindow::clipbaord_changed.

seems right.

Let's see if @Memory383  can still encounter this crash after the commit https://github.com/xiaoyifang/goldendict/commit/84798c56eaed463f07ce4678e3dfbac7b8f14daf

**Comment by xiaoyifang** [comment]
> `this` and nullptr is not related to possible crashes too, because Scanpopup won't be destroyed at all unless the app quit

you are right . it will not lead to crash .   It is only related to memory leak I think.

**Comment by Memory383** [comment]
I did a fresh installation after the latest commit. 

GD still crashes every few hours, but it can be used. _It is not a critical issue._ 

Let's keep this issue open. Maybe someone else with a background in computer science could help us pinpoint the problem.

**Comment by shenlebantongying** [comment]
Crash after long-running is hard to figure out exactly why.

This is the time, my friend. You just learned how to get stack traces from systemd, and you use Manjaro which is arch-based, and you are a hardcore GD user. The combination cannot be better.

Let's find out. You need to switch to a debug build, because release build omitted some info and render a lot of things in stacktraces as `n/a`.

```
sudo pacman -S base-devel

# uninstall the existing GD

git clone https://aur.archlinux.org/goldendict-webengine-git.git
cd goldendict-webengine-git

# open PKGBUILD, and change the CONFIG+=release to CONFIG+=debug inside the build section.

makepkg --syncdeps --install 
```

 It will be slower but contains all the info.

**Comment by Memory383** [comment]
The issue of crashes seems solved after the latest commit : 9079e2125d0e2ab9e5d521b10e48355951dca126 

I installed the “debug” version after that commit and everything works perfectly ! 

Thanks to my friends Xiao and @shenlebantongying !

**Comment by shenlebantongying** [comment]
@Memory383 Please don't fool me, that commit has nothing todo with "crashes every few hours". The code in the commit only matters for startup. The point of debug build is to have better stacktrace to know exact point of crash.

**Comment by Memory383** [comment]
@shenlebantongying   Sorry. My message was confusing.  I tried the "debug" and the "release" builds and there were no crashes in several hours. The most surprising improvement is that the GUI loads correctly now after launching the software, and it does not crash.

 Before the commits 84798c56eaed463f07ce4678e3dfbac7b8f14daf and 9079e2125d0e2ab9e5d521b10e48355951dca126 GD would sometimes crash after startup. Other times, it would crash hours later.

@xiaoyifang  I used the "debug" version and everything was perfect. The only "issue" is that queries felt a bit slower. Then, I switched to the "release" version, and it also works flawlessly. 

-----
I guess that the crashes had a relationship with the clipboard because I use simultaneously 4 software that track the clipboard manager:
1) KDE Native Clipboard Manager
2) CopyQ
3) VocabSieve
4) GoldenDict

All of them track the primary mouse selection on my PC and crashes might had a relation with them. Anyway, the problem seems solved. Thanks again!
